Zoom bar

hlyang397 opened the pull request:
### Updates
For #380 
- add new ZoomBar component

### Demo screenshot or recording
![image](https://user-images.githubusercontent.com/59426533/86451328-56866080-bd4d-11ea-8d84-a1814e2ce5e5.png)
****
### Review checklist (for reviewers only)
- [ ] Demos all features
- [ ] Documented/annotated
- [ ] Matches UI/UX specs
- [ ] Meets the [code style guide](https://github.com/carbon-design-system/carbon-charts/wiki/Code-style-guide)
- [ ] Accessible
- [ ] Mobile first (responsive)
- [ ] RTL support (bidirectional text)
- [ ] Performant (limited bloat)

theiliad commented:
converting this PR into a draft

hlyang397 commented:
> Looking great :hlyang397 💯
> One thing I noticed is that the zooming doesn't happen continuously (as you're moving the handles), that is probably a key aspect of the experience to have the graph update in real-time.

@theiliad Thanks for your review.
We have many new commits in past few days, all comments are welcome, thanks.

theiliad commented:
Hi, looks great! 💯 

Couple of pieces of feedback:
1. we should enable the feature in more charts just to validate the functionality and make sure that it works everywhere
2. I'm seeing a bug sometimes where the axes don't resize properly
![image](https://user-images.githubusercontent.com/14989804/84798900-7ede3300-afc9-11ea-9dfb-cb3d72c9547c.png)

3. as soon as I move the handles on both ends, they resize and become taller, they should probably stay the same height
![image](https://user-images.githubusercontent.com/14989804/84799006-a1704c00-afc9-11ea-9658-2b85f9967f78.png)

4. Not sure if this is a bug but when I go any further to the left from this point, the graph takes no changes
![image](https://user-images.githubusercontent.com/14989804/84799225-f7dd8a80-afc9-11ea-9371-a4e6ea68eac7.png)

theiliad commented:
question, do we have a way of setting the range of the zoomed area through options? e.g. if you were rendering a graph with many points and wanted it initially to zoom into an area and only show 10-20 points?

theiliad commented:
TODOs that I can see so far:
- click&drag behavior on the graph should trigger panning not zooming
- shift-clicking should result in zooming
- handle toolbars aren't in yet
- we need the non-highlighted areas in the zoom-bar to be darker in dark themes and lighter in light themes (I'd just draw 2 rectangles over both sides)
- `addSpaceOnEdges` is doing nothing now when the zoom bar shows up
- chart renders incorrectly initially and once you interact with the zoom bar it re-renders correctly
![image](https://user-images.githubusercontent.com/14989804/85326982-29f45e00-b49c-11ea-8d33-9d5834eba1c3.png)

- tooltips are no longer working on the graphs
- handles have the wrong height
![image](https://user-images.githubusercontent.com/14989804/85326969-22cd5000-b49c-11ea-9dbb-0c894cf21f33.png)

theiliad commented:
@scottdickerson

hlyang397 commented:
@theiliad,

Thanks for your review comment.

>click&drag behavior on the graph should trigger panning not zooming
I think zooming makes more sense than panning when user selects an area.
Are you talking about the dragging event on the graph ? We didn't implement it yet.


>shift-clicking should result in zooming
>handle toolbars aren't in yet
We will include these features in next phase.

> handles have the wrong height
The handle height is overwritten by d3.js brush library. I would put this as low priority before I can find the solution.


For rest problems, I have created issues in our tool.
Thanks.

hlyang397 commented:
@theiliad,
Please help to review the latest changes.

There are some issues still in progress, and may need more discussion.

1. We think click & drag on the graph should trigger zooming not panning.
2. Space on edges makes sense in some charts (ex: bar, bubble), but looks weird in the others(ex: line).
With zoombar enabled, we may need to handle them accordingly.
3. Tooltips are covered by the chart brush. Waiting for further investigation.
4. Handle toolbar and shift-clicking zooming will be included in next phase (PR).

All comments are welcome, thanks.

theiliad commented:
> @theiliad,
> Please help to review the latest changes.
> 
> There are some issues still in progress, and may need more discussion.
> 
> 1. We think click & drag on the graph should trigger zooming not panning.
@designertyler 

> 2. Space on edges makes sense in some charts (ex: bar, bubble), but looks weird in the others(ex: line).
>    With zoombar enabled, we may need to handle them accordingly.
we can't override what the user defines in options, if they add space on the edges, we need to show it.

> 3. Tooltips are covered by the chart brush. Waiting for further investigation.
![image](https://user-images.githubusercontent.com/14989804/86019942-442ac080-b9f5-11ea-9dd8-a66af1be81a7.png)

would something like:
```css
react.overlay:hover {
	pointer-events: none;
}
```

work?

we've gotta find a way to show tooltips on hover

> 4. Handle toolbar and shift-clicking zooming will be included in next phase (PR).
👍 

> 
> All comments are welcome, thanks.

hlyang397 commented:
@theiliad ,
Thanks for your comments.
2. That makes sense to me. I will try to make the `addSpaceOnEdges` option working with zoombar.
3. Setting `pointer-events: none;` breaks the brush. I fix that by moving brush component layer to back in this commit.
https://github.com/carbon-design-system/carbon-charts/pull/649/commits/60990b94af012bcbb34d93ec9512445e900d7cc3

theiliad commented:
> @theiliad ,
> Thanks for your comments.
> 2. That makes sense to me. I will try to make the `addSpaceOnEdges` option working with zoombar.
> 3. Setting `pointer-events: none;` breaks the brush. I fix that by moving brush component layer to back in this commit.
> [60990b9](https://github.com/carbon-design-system/carbon-charts/commit/60990b94af012bcbb34d93ec9512445e900d7cc3)

So the singular tooltip works now but not the multiline tooltip. I think we might have to get rid of the brush overlay and release events such as `grid-backdrop-drag` etc... and show the zoom area by catching that

hlyang397 commented:
@theiliad ,
> So the singular tooltip works now but not the multiline tooltip. I think we might have to get rid of the brush overlay and release events such as `grid-backdrop-drag` etc... and show the zoom area by catching that
Could you give me an example of multiline tooltip? Thanks.

theiliad commented:
![image](https://user-images.githubusercontent.com/14989804/86164509-de673300-badf-11ea-858c-404f3ef265b7.png)

https://carbon-design-system.github.io/carbon-charts/?path=/story/line--line-time-series

hlyang397 commented:
@theiliad,
1. The `addSpaceOnEdges` is working in latest commit.
2. The multiline tooltip is in progress.

theiliad commented:
> @theiliad,
> 
> 1. The `addSpaceOnEdges` is working in latest commit. Please let me know if any comment.
> 2. The fix of multiline tooltip is in progress.
>    Thanks.

Thanks. A couple of points that I'm currently noticing:
1. The overlay seems to be off by a bit
![image](https://user-images.githubusercontent.com/14989804/86319470-9f7dce00-bc02-11ea-91b6-4aa10069f946.png)

2. The handles seem to always snap on points where you can only see half of a bar on the edge
![image](https://user-images.githubusercontent.com/14989804/86319551-c20fe700-bc02-11ea-983f-ee4e9bff53be.png)

Also adding @designertyler & @jeanservaas here

hlyang397 commented:
@theiliad ,
Both issues are addressed in latest commits, please help to review again, thanks.
All planned functionalities are completed.
I will change this draft to PR after some code refactorings, please let me know if any comment, thanks.

theiliad commented:
> @theiliad ,
> Both issues are addressed in latest commits, please help to review again, thanks.
> All planned functionalities are completed.
> I will change this draft to PR after some code refactorings, please let me know if any comment, thanks.

ok thanks, I still have to review the code thoroughly. Hoping to do that after you've marked it ready for review

sstone2423 commented:
Hello, this looks great so far. I'm testing this locally with our `carbon-addons-iot-react` package now as we'll be using it in our internal application once its fully available. I'll let you know if I spot any bugs, but the storybook demos looks solid so far

theiliad commented on packages/core/src/components/graphs/scatter.ts:
```diff
@@ -64,6 +77,9 @@ export class Scatter extends Component {
 		}
 
+		// filter out of domain data
+		scatterData = this.filterOutOfDomain(scatterData);
```
is this something that needs to live in the `Scatter` component?

hlyang397 replied:
Yes, we may see partial circle at the edge in some charts.
![image](https://user-images.githubusercontent.com/59426533/87134104-c99e5280-c2ca-11ea-99ce-4bc52ebec3d1.png)

theiliad commented on packages/core/src/components/graphs/line.ts:
```diff
@@ -145,7 +146,7 @@ export class Line extends Component {
 	handleLegendMouseOut = (event: CustomEvent) => {
 		this.parent
-			.selectAll("path.line")
+			.selectAll(`g#coverClip`)
```
what does this do?

hlyang397 replied:
Good catch! These should be deleted.
Thanks.

theiliad commented on packages/core/src/components/component.ts:
```diff
@@ -90,10 +90,28 @@ export class Component {
+				this.type === "grouped-bar" ||
+				this.type === "simple-bar" ||
+				this.type === "stacked-bar" ||
+				this.type === "scatter-stacked"
```
yes this doesn't make sense

hlyang397 replied:
Thanks for your comment.
I use a flag in getContainerSVG() to avoid this problem.

hlyang397 commented on packages/core/src/configuration.ts:
```diff
@@ -154,6 +155,17 @@ export const timeScale: TimeScaleOptions = {
+ */
+export const zoomBar: ZoomBarOptions = {
+	enabled: false,
+	initZoomDomain: undefined,
```
Good point, we don't need to.

hlyang397 commented on packages/core/src/components/axes/axis.ts:
```diff
@@ -402,10 +408,10 @@ export class Axis extends Component {
-				rotateTicks = estimatedTickSize < minTickSize;
+				rotateTicks = isTimeScaleType
+					? estimatedTickSize < minTickSize * 2 // datetime tick could be very long
+					: estimatedTickSize < minTickSize;
```
If it's timeScaleType, we need a different threshold to rotate ticks because time labels are usually very long and overlapped easily.

hlyang397 commented on packages/core/src/charts/area-stacked.ts:
```diff
@@ -34,8 +35,9 @@ export class StackedAreaChart extends AxisChart {
 	getComponents() {
 		// Specify what to render inside the graph-frame
-		const graphFrameComponents = [
+		const graphFrameComponents: any[] = [
```
It works, and it could be `Component[]`, but not necessary.
We just follow the interface of `getAxisChartComponents(graphFrameComponents: any[])`.
If you think this is a problem, I would recommend to create another PR for this type change.

hlyang397 commented on packages/core/demo/data/zoom-bar.ts:
```diff
@@ -0,0 +1,95 @@
+const defaultZoomBarOptions = {
+	enabled: true,
+	initZoomDomain: undefined,
+	selectionStart: selectionStartFun,
```
@theiliad,
These callbacks are used by external user, not inside the library.
If we like to dispatch events to library user, we need to provide the API for user to get Events service, right ?
Please let me know if there is anything I missed here.

hlyang397 commented on packages/core/src/chart.ts:
```diff
@@ -110,7 +111,7 @@ export class Chart {
 		this.services.domUtils.getHolder().remove();
 
-		this.model.set({ destroyed: true }, true);
+		this.model.set({ destroyed: true }, { skipUpdate: true });
```
I am not pretty sure.
I guess we should set `destroyed: true` after every component is destroyed ?
The `skipUpdate` here is just to make sure that it won't trigger the update.
But at this point, every component should be destroyed.

BTW, I didn't try to change logic here, the update of this line is just to follow the mode.set() interface change.

hlyang397 commented on packages/core/src/charts/area-stacked.ts:
```diff
@@ -34,8 +35,9 @@ export class StackedAreaChart extends AxisChart {
-		const graphFrameComponents = [
+		const graphFrameComponents: any[] = [
 			new TwoDimensionalAxes(this.model, this.services),
+			new Cover(this.model, this.services),
```
`Cover` is changed to `ChartClip`.
This is used to clip chart graphs out of bound.
It's necessary for zoom in/out behavior.

hlyang397 commented on packages/core/src/components/axes/brush.ts:
```diff
@@ -0,0 +1,136 @@
+import { event } from "d3-selection";
+import { scaleTime } from "d3-scale";
+
+export class Brush extends Component {
```
Change `Brush` to `ChartBrush`.
This class is used for handle brush events in chart.
It's only used by zoom in for now, but it could be improved for other brush events in the future.

hlyang397 commented on packages/core/src/model.ts:
```diff
@@ -267,11 +309,11 @@ export class ChartModel {
 	}
 
-	set(newState: any, skipUpdate = false) {
+	set(newState: any, configs?: any) {
```
1. All callers are updated to new interface. Please refer to the usage report below.
2. The set function is updated with a default value.
```
    Usage in function call  (10 usages found)
        carbon-charts  (10 usages found)
            packages/core/src  (5 usages found)
                chart.ts  (2 usages found)
                    Chart  (2 usages found)
                        destroy  (1 usage found)
                            114 this.model.set({ destroyed: true }, { skipUpdate: true });
                        init  (1 usage found)
                            40 this.model.set({ holder }, { skipUpdate: true });
                model.ts  (3 usages found)
                    ChartModel  (3 usages found)
                        setData  (1 usage found)
                            166 this.set({
                        setOptions  (1 usage found)
                            336 this.set({
                        toggleDataLabel  (1 usage found)
                            430 this.set({
            packages/core/src/components/axes  (5 usages found)
                chart-brush.ts  (2 usages found)
                    ChartBrush  (2 usages found)
                        brushed  (1 usage found)
                            109 this.model.set(
                        render  (1 usage found)
                            39 this.model.set({ zoomDomain: zoomDomain }, { animate: false });
                two-dimensional-axes.ts  (1 usage found)
                    TwoDimensionalAxes  (1 usage found)
                        render  (1 usage found)
                            119 this.model.set({ axesMargins: this.margins }, { skipUpdate: true });
                zoom-bar.ts  (2 usages found)
                    ZoomBar  (2 usages found)
                        brushed  (1 usage found)
                            242 this.model.set({ zoomDomain: newDomain }, { animate: false });
                        init  (1 usage found)
                            43 this.model.set(
```

hlyang397 commented on packages/core/src/services/scales-cartesian.ts:
```diff
@@ -157,33 +156,65 @@ export class CartesianScales extends Service {
+		axisPosition: AxisPositions,
+		datum: any,
+		index?: number
+	) {
```
In zoom bar, we will have another scale different from the chart scale, so we improve this function to improve reusability.

hlyang397 commented on packages/core/src/components/axes/brush.ts:
```diff
@@ -0,0 +1,136 @@
+
+		const { cartesianScales } = this.services;
+		const mainXAxisPosition = cartesianScales.getMainXAxisPosition();
+		const mainXScaleType = cartesianScales.getScaleTypeByPosition(
```
Good point!

hlyang397 commented on packages/core/src/components/axes/axis.ts:
```diff
@@ -106,6 +106,12 @@ export class Axis extends Component {
+		const zoomDomain = this.model.get("zoomDomain");
+		if (zoomDomain && axisPosition === AxisPositions.BOTTOM) {
+			scale.domain(zoomDomain);
+		}
```
Good point! Checking more options in `axis-chart.ts`.

hlyang397 commented on packages/core/src/interfaces/events.ts:
```diff
@@ -17,6 +17,13 @@ export enum Model {
+ */
+export enum ZoomBar {
+	UPDATE = "zoom-bar-update"
+}
```
Let's have more discussion in this thread.
https://github.com/carbon-design-system/carbon-charts/pull/649#discussion_r452046457

hlyang397 commented on packages/core/src/interfaces/components.ts:
```diff
@@ -161,3 +161,31 @@ export interface BarOptions {
+	/**
+	 * a function to handle selection start event
+	 */
+	selectionStart?: Function;
+	/**
+	 * a function to handle selection in progress event
+	 */
+	selectionInProgress?: Function;
+	/**
+	 * a function to handle selection end event
+	 */
+	selectionEnd?: Function;
```
Let's discuss in this thread.
https://github.com/carbon-design-system/carbon-charts/pull/649#discussion_r452046457

hlyang397 commented on packages/core/src/model.ts:
```diff
@@ -38,7 +39,48 @@ export class ChartModel {
 	}
+	// get display data for zoom bar
+	// basically it's sum of value grouped by time
+	getZoomBarData() {
```
getZoomBarData() and getDefaultZoomBarDomain() are shared to ZoomBar and ChartBrush.
Please let me know if any idea how we could share these functions better. Thanks.

hlyang397 commented on packages/core/src/services/essentials/dom-utils.ts:
```diff
@@ -149,6 +149,30 @@ export class DOMUtils extends Service {
 		return selection;
 	}
 
+	static appendOrSelectForAxisChart(parent, query) {
```
Thanks for catching this. This is not a good approach.
I have remove this function in latest commit.

hlyang397 commented:
@theiliad ,
I believe I have addressed all your comments in latest commits except the external callbacks.
I can't use event handler because the callbacks are for external library caller.

Please help to review this PR again, thanks.

designertyler commented:
- [ ] The x-axis labels change orientation when the scrolling. Can we lock in the orientation to avoid this jumping?

![Jul-10-2020 11-36-02](https://user-images.githubusercontent.com/19270502/87177931-0bff6980-c2a2-11ea-9089-3c271259d527.gif)


- [ ] Based on the specs in https://github.com/carbon-design-system/carbon-charts/issues/380 the drag selection should be above the visualization
![image](https://user-images.githubusercontent.com/19270502/87178776-503f3980-c2a3-11ea-89bd-fc15127bb8f5.png)

- [ ] The end drag point is dashed in the specs.
![image](https://user-images.githubusercontent.com/19270502/87179261-0dca2c80-c2a4-11ea-94bd-1fe551cd1111.png)


- [ ] Hover states for the zoom handles are missing
![image](https://user-images.githubusercontent.com/19270502/87179389-48cc6000-c2a4-11ea-9c48-47599a5792de.png)


- [ ] The specs show the click/drag behavior is reserved for panning and the shift/click is used for zooming. Maybe the requirements were changes since then?
![image](https://user-images.githubusercontent.com/19270502/87179596-a1036200-c2a4-11ea-8131-7ff2eac88bb2.png)

hlyang397 commented:
@theiliad ,
Zoombar now dispatches events instead of function callbacks per our discussion.
Please help to review this PR again, thanks.

hlyang397 commented:
@designertyler 
Thanks for your review.
1. The label rotation depends on the tick count and label size, and these could change while moving zoom bar handle. If we lock the rotation, it could cause some overlaps or unnecessary rotation.
2. The drag selection is behind those graphs(line, area, etc) to allow some mouseover events (scatter tooltip, ruler multiple tooltip) are still working. I am still thinking how to fix this if possible.
3. The end drag point dash is fixed in last commits.
4. Now I use SVG element title as tooltip. Is it acceptable for now? The customized tooltip may need more efforts.
5. Our team believes that dragging to zoom in makes more sense. Not sure if you agree to change the spec ?
Thanks.

theiliad commented:
> @designertyler
> Thanks for your review.
> 
> 1. The label rotation depends on the tick count and label size, and these could change while moving zoom bar handle. If we lock the rotation, it could cause some overlaps or unnecessary rotation.
> 2. The drag selection is behind those graphs(line, area, etc) to allow some mouseover events (scatter tooltip, ruler multiple tooltip) are still working. I am still thinking how to fix this if possible.
> 3. The end drag point dash is fixed in last commits.
> 4. Now I use SVG element title as tooltip. Is it acceptable for now? The customized tooltip may need more efforts.
> 5. Our team believes that dragging to zoom in makes more sense. Not sure if you agree to change the spec ?
>    Thanks.

4. I'm updating tooltips right now so it'll be easier to add tooltips to handles, #695

theiliad commented:
> 4. Now I use SVG element title as tooltip. Is it acceptable for now? The customized tooltip may need more efforts.

I'm updating tooltips right now so it'll be easier to add tooltips to handles, #695

hlyang397 commented:
> > 1. Now I use SVG element title as tooltip. Is it acceptable for now? The customized tooltip may need more efforts.
> 
> I'm updating tooltips right now so it'll be easier to add tooltips to handles, #695

@theiliad Thanks for your information. That's helpful.
But I hope this PR won't be blocked by #695. We could create another issue for the tooltip upgrade.
Or this PR could last forever.

hlyang397 commented:
@designertyler

1. The label rotation is set to always true while zoom domain is changing. Hope it answers your question.
2. The drag selection is now in front of those graphs(line, area, etc).

hlyang397 commented:
@theiliad @designertyler @natashadecoste 
Do you have any update to the review ?
We would like to have this PR merged as soon as possible.
Please let me know if anything I need to change.
Thanks.

theiliad commented:
> @theiliad @designertyler @natashadecoste
> Do you have any update to the review ?
> We would like to have this PR merged as soon as possible.
> Please let me know if anything I need to change.
> Thanks.

Hi, I'd have to add in the newly designed pieces of the UI and make things more scalable. I'll let you know

natashadecoste commented:
right now the `rect.zoom-bg` is going outside the chart container 
![image](https://user-images.githubusercontent.com/14351335/88956308-302fe400-d26b-11ea-9a3c-80bf3c99aebe.png)
